src/simulation/LoadGenerator.cpp

```diff
@@ -999,37 +1677,54 @@ LoadGenerator::checkAccountSynced(Application& app, bool isCreate)
 }
 
 void
-LoadGenerator::waitTillComplete(bool isCreate)
+LoadGenerator::waitTillComplete(GeneratedLoadConfig cfg)
 {
     if (!mLoadTimer)
     {
         mLoadTimer = std::make_unique<VirtualTimer>(mApp.getClock());
     }
     vector<TestAccountPtr> inconsistencies;
-    inconsistencies = checkAccountSynced(mApp, isCreate);
+    inconsistencies = checkAccountSynced(mApp, cfg.isCreate());
+    auto sorobanInconsistencies = checkSorobanStateSynced(mApp, cfg);
 
-    if (inconsistencies.empty())
+    // If there are no inconsistencies and we have generated all load, finish
+    if (inconsistencies.empty() && sorobanInconsistencies.empty() &&
+        cfg.isDone())
     {
         CLOG_INFO(LoadGen, "Load generation complete.");
         mLoadgenComplete.Mark();
-        reset();
+        reset(/*resetSoroban=*/false);
         return;
     }
-    else
+    // If we have an inconsistency, reset the timer and wait for another ledger
+    else if (!inconsistencies.empty() || !sorobanInconsistencies.empty())
     {
         if (++mWaitTillCompleteForLedgers >= TIMEOUT_NUM_LEDGERS)
         {
             CLOG_INFO(LoadGen, "Load generation failed.");
             mLoadgenFail.Mark();
-            reset();
+            reset(/*resetSoroban=*/!sorobanInconsistencies.empty());
```

Comment:
nit: it seems that resetting soroban state doesn't actually belong to `reset`, which resets all state between runs. probably better to have a dedicated function for soroban state reset, this way you can avoid duplication on line 322 as well.